### ISSUE TITLE
[Phase K] Benchmark ModernBERT dLLM SFT performance: unturtle vs dllm

### ISSUE THREAD

**Issue by nishide-dev** [nishide-dev/unturtle-legacy#104]
## 背景

このプロジェクトの目的はdLLM学習を最適化すること。
unturtleの最適化（Triton kernel、双方向attention、LoRAパッチ）が
既存フレームワーク (dllm/MDLM) に対してVRAM使用量・学習時間・lossでどれだけ改善しているか、
実測値で比較検証したい。

## 目的

`answerdotai/ModernBERT-base`（145Mパラメータ）を同一データ・同一ハイパーパラメータで学習し、
unturtle と dllm の実測値を比較する。

## Acceptance Criteria

- [ ] `notebooks/benchmark_modernbert_unturtle.py` で unturtle側ベンチマーク実行
- [ ] `notebooks/benchmark_modernbert_dllm.py` で dllm側ベンチマーク実行（別venv必要）
- [ ] `notebooks/benchmark_modernbert_compare.py` で比較表生成
- [ ] 以下のメトリクスを収集・記録:
  - Train loss (avg / first 50 / last 50)
  - Elapsed time
  - Peak VRAM
  - Steps/sec
- [ ] 結果をCLAUDE.mdのベンチマークセクションに記録

## 実行方法

```bash
# unturtle
source .venv/bin/activate
python notebooks/benchmark_modernbert_unturtle.py

# dllm (別venv)
.venvDllm/bin/python notebooks/benchmark_modernbert_dllm.py

# 比較
python notebooks/benchmark_modernbert_compare.py
```

## 実装内容

| スクリプト | 用途 |
|-----------|------|
| `notebooks/benchmark_modernbert_unturtle.py` | unturtle DiffusionTrainer ベンチマーク |
| `notebooks/benchmark_modernbert_dllm.py` | dllm MDLMTrainer ベンチマーク (別venv用) |
| `notebooks/benchmark_modernbert_compare.py` | 結果比較・レポート生成 |

## 注意事項

- dllmは transformers 4.57.0 依存のため `.venvDllm/` で隔離
- `FastDiffusionModel.from_pretrained("answerdotai/ModernBERT-base", model_class=DiffusionModernBertForMaskedLM, ...)` 経由でロード
- LoRA target: `["Wqkv", "Wo"]`

**Comment by nishide-dev** [comment]
## ベンチマーク結果 (2026-04-05, RTX PRO 6000 Blackwell Max-Q)

**セットアップ**: ModernBERT-base (0.150B) / batch=4, max_len=512, LoRA r=16, α=32, lr=1e-5
**データ**: allenai/tulu-3-sft-mixture (train=1,989, eval=170) / 2 epochs

| メトリック | **unturtle** | **dllm** | 倍率 |
|-----------:|:------------:|:--------:|:----:|
| steps/sec | **6.39** | 2.90 | **2.2x** |
| samples/sec | **25.5** | 11.6 | **2.2x** |
| 所要時間 | **2.6 min** | 5.8 min | **2.2x** |
| peak VRAM | **1.55 GB** | 2.85 GB | **0.54x** |
| train_loss (avg) | 2.09 | 2.01 | — |
| eval_loss (epoch 2) | 2.11 | 2.32 | — |

### 要点
- **学習速度が 2.2x 高速** (Triton カーネルによる損失最適化 + LoRA fast path)
- **VRAM が 46% 削減** (1.55 vs 2.85 GB)
- eval_loss も 2.11 vs 2.32 で unturtle が優位 (正規化方式の違いにより絶対値は直接比較不可)

### dllm 側でハマった点
- PyTorch 2.6.0+cu124 は Blackwell (sm_120) 非対応 → nightly cu128 版に差し替えで解決
- Multi-GPU DataParallel + `torch.compile` (compiled_mlp) の競合 → `CUDA_VISIBLE_DEVICES=0` でシングルGPU化
- ModernBERT tokenizer に chat_template 無し → 手動フラッテン方式に切り替え
